streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala

```diff
@@ -488,7 +493,28 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
           registerReceiver(streamId, typ, host, executorId, receiverEndpoint, context.senderAddress)
         context.reply(successful)
       case AddBlock(receivedBlockInfo) =>
-        context.reply(addBlock(receivedBlockInfo))
+        if (WriteAheadLogUtils.isBatchingEnabled(ssc.conf, isDriver = true)) {
+          val f = Future {
+            synchronized {
+              if (active) {
+                addBlock(receivedBlockInfo)
+              } else {
+                throw new IllegalStateException("Receiver Tracker Endpoint shutdown.")
+              }
+            }
+          }(walBatchingThreadPool)
+          f.onComplete { result =>
+            synchronized {
+              if (active) {
+                context.reply(result.get)
+              } else {
+                context.reply(new IllegalStateException("Receiver Tracker Endpoint shutdown."))
```

Comment:
You are returning an exception here ... are you sure that the other side can receiver exceptions at all? I think it only receives boolean - https://github.com/apache/spark/blob/master/streaming/src/main/scala/org/apache/spark/streaming/receiver/ReceiverSupervisorImpl.scala#L161